### PR TITLE
Supports quantifiers in function contracts

### PR DESCRIPTION
**Pull request by ArenBabikian**
This PR adds support for quantified expressions in function contracts. The current implementation handles cases where a function contract is only composed of a single quantified expression. Additionally, nested quantified expressions are not yet supported.

This PR might be a continuation of #2278. 
Additionally, #5968 is a follow-up to this PR.

---

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: regression/contracts/function_check_02/test.desc [hunk 1]
```diff
@@ -1,6 +1,6 @@
 KNOWNBUG
 main.c
---check-code-contracts
+--enforce-all-contracts
```
Comment (feliperodri) [inline regression/contracts/function_check_02/test.desc:3]: If it's still a known bug, even with `--enforce-all-contracts`, does the description is still accurate? I believe this change doesn't belong to this PR.

File: regression/contracts/quantifiers-exists-ensures-01/main.c [hunk 2]
```diff
@@ -0,0 +1,18 @@
+int f1(int *arr) __CPROVER_ensures(__CPROVER_exists {
+  int i;
+  (0 <= i && i < 10) && arr[i] == i
+})
```
Comment (feliperodri) [inline regression/contracts/quantifiers-exists-ensures-01/main.c]: We can only use quantifiers in positive context, correct? Is there anywhere where this is documented for now? Could we add a regression test to expose this problem using contracts (e.g., using in requires with enforce).

File: regression/contracts/quantifiers-exists-ensures-01/test.desc [hunk 3]
```diff
@@ -0,0 +1,10 @@
+CORE
+main.c
+--enforce-all-contracts
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+--
+--
+Verification:
+This test asserts the postconditions of f1.
```
Comment (feliperodri) [inline regression/contracts/quantifiers-exists-ensures-01/test.desc]: The purpose of this test is to ensure we can safety use `__CPROVER_exists` in require clauses. The expected behavior is that using `--enforce-all-contracts` flag will transform requires into assertions and the verification remains sound when using `__CPROVER_exists`. Could you update the test description for this test (same for all the others). Also, `test.desc` lines should be < 80 characters.

File: regression/contracts/quantifiers-exists-ensures-02/test.desc [hunk 4]
```diff
@@ -0,0 +1,10 @@
+CORE
+main.c
+--enforce-all-contracts
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
+--
+Verification:
+This test asserts the postconditions of f1.
```
Comment (feliperodri) [inline regression/contracts/quantifiers-exists-ensures-02/test.desc]: Update description as described above.

File: regression/contracts/quantifiers-exists-ensures-01/main.c (hunk 2 above)
Comment (feliperodri) [reply regression/contracts/quantifiers-exists-ensures-01/main.c]: Why not `(0 <= i && i < 10) ==> arr[i] == i`?

File: regression/contracts/quantifiers-exists-ensures-02/main.c [hunk 5]
```diff
@@ -0,0 +1,18 @@
+int f1(int *arr) __CPROVER_ensures(__CPROVER_exists {
+  int i;
+  (0 <= i && i < 10) && arr[i] != 0
```
Comment (feliperodri) [inline regression/contracts/quantifiers-exists-ensures-02/main.c]: Why not `(0 <= i && i < 10) ==> arr[i] != 0`?

File: regression/contracts/quantifiers-exists-requires-01/main.c [hunk 6]
```diff
@@ -0,0 +1,13 @@
+int f1(int *arr) __CPROVER_requires(__CPROVER_exists {
+  int i;
+  (0 <= i && i < 10) && arr[i] == 4
```
Comment (feliperodri) [inline regression/contracts/quantifiers-exists-requires-01/main.c]: Why not `(0 <= i && i < 10) ==> arr[i] == 4`?

File: regression/contracts/quantifiers-exists-requires-01/test.desc [hunk 7]
```diff
@@ -0,0 +1,10 @@
+CORE
+main.c
+--replace-all-calls-with-contracts
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+--
+--
+Verification:
+This test asserts the preconditions of f1.
```
Comment (feliperodri) [inline regression/contracts/quantifiers-exists-requires-01/test.desc]: Update description.

File: regression/contracts/quantifiers-exists-requires-02/main.c [hunk 8]
```diff
@@ -0,0 +1,13 @@
+int f1(int *arr) __CPROVER_requires(__CPROVER_exists {
+  int i;
+  (0 <= i && i < 10) && arr[i] == 1
```
Comment (feliperodri) [inline regression/contracts/quantifiers-exists-requires-02/main.c]: Why not `(0 <= i && i < 10) ==> arr[i] == 1`?

File: regression/contracts/quantifiers-exists-requires-02/test.desc [hunk 9]
```diff
@@ -0,0 +1,10 @@
+CORE
+main.c
+--replace-all-calls-with-contracts
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
+--
+Verification:
+This test asserts the preconditions of f1.
```
Comment (feliperodri) [inline regression/contracts/quantifiers-exists-requires-02/test.desc]: Update description.

File: regression/contracts/quantifiers-forall-ensures-01/test.desc [hunk 10]
```diff
@@ -0,0 +1,10 @@
+CORE
+main.c
+--replace-all-calls-with-contracts
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+--
+--
+Verification:
+This test asserts the preconditions of f1.
```
Comment (feliperodri) [inline regression/contracts/quantifiers-forall-ensures-01/test.desc]: Update description.

File: regression/contracts/quantifiers-forall-ensures-02/test.desc [hunk 11]
```diff
@@ -0,0 +1,10 @@
+CORE
+main.c
+--enforce-all-contracts
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
+--
+Verification:
+This test asserts the postconditions of f1.
```
Comment (feliperodri) [inline regression/contracts/quantifiers-forall-ensures-02/test.desc]: Update description.

File: regression/contracts/quantifiers-forall-requires-01/test.desc [hunk 12]
```diff
@@ -0,0 +1,10 @@
+CORE
+main.c
+--replace-all-calls-with-contracts
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+--
+--
+Verification:
+This test asserts the preconditions of f1.
```
Comment (feliperodri) [inline regression/contracts/quantifiers-forall-requires-01/test.desc]: Update description.

File: regression/contracts/quantifiers-forall-requires-02/test.desc [hunk 13]
```diff
@@ -0,0 +1,10 @@
+CORE
+main.c
+--replace-all-calls-with-contracts
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
+--
+Verification:
+This test asserts the preconditions of f1.
```
Comment (feliperodri) [inline regression/contracts/quantifiers-forall-requires-02/test.desc]: Update description.

File: src/goto-instrument/code_contracts.cpp [hunk 14]
```diff
@@ -163,6 +163,41 @@ bool code_contractst::has_contract(const irep_idt fun_name)
...
+  replace_symbolt &replace,
+  irep_idt mode)
+{
+  // If the expression is a quantified expression, this function adds
+  // the quantified variable to the symbol table and to the expression map
+
+  // TODO Currently only works if the contract contains only a single
+  // quantified formula
+  // i.e. (1) the top-level element is a quantifier formula
+  // and (2) there are no inner quantifier formulae
```
Comment (feliperodri) [inline src/goto-instrument/code_contracts.cpp:177]: Can we turn this into a `INVARIANT(...)`? Then we should add a GitHub issue to track that.

File: src/goto-instrument/code_contracts.cpp [hunk 15]
```diff
@@ -163,6 +163,41 @@ bool code_contractst::has_contract(const irep_idt fun_name)
...
+  if(expression.id() == ID_exists || expression.id() == ID_forall)
```
Comment (feliperodri) [inline src/goto-instrument/code_contracts.cpp:180]: Could we also turn this into an invariant instead of an if statement (e.g., `INVARIANT(expression.id() == ID_exists || expression.id() == ID_forall)`)? I think this function will only be called for these two expressions.

File: src/goto-instrument/code_contracts.cpp [hunk 16]
```diff
@@ -149,7 +149,7 @@ static void check_apply_invariants(
 
   // change the back edge into assume(false) or assume(guard)
   loop_end->targets.clear();
-  loop_end->type=ASSUME;
+  loop_end->type = ASSUME;
```
Comment (feliperodri) [inline src/goto-instrument/code_contracts.cpp]: This change doesn't belong here.

File: src/goto-instrument/code_contracts.cpp [hunk 17]
```diff
@@ -230,7 +265,7 @@ bool code_contractst::apply_function_contract(
   }
 
   // Replace formal parameters
-  code_function_callt::argumentst::const_iterator a_it=
+  code_function_callt::argumentst::const_iterator a_it =
```
Comment (feliperodri) [inline src/goto-instrument/code_contracts.cpp]: This change doesn't belong here.

File: src/goto-instrument/code_contracts.cpp [hunk 18]
```diff
@@ -745,7 +785,7 @@ void code_contractst::add_contract_check(
                        .symbol_expr();
     check.add(goto_programt::make_decl(r, skip->source_location));
 
-    call.lhs()=r;
+    call.lhs() = r;
```
Comment (feliperodri) [inline src/goto-instrument/code_contracts.cpp]: This change doesn't belong here.

File: src/goto-instrument/code_contracts.cpp [hunk 19]
```diff
@@ -440,38 +480,38 @@ void code_contractst::instrument_call_statement(
...
+    assigns_clauset called_assigns_clause(
+      called_assigns, *this, function_id, log);
+    exprt compatible =
+      assigns_clause.compatible_expression(called_assigns_clause);
+    goto_programt alias_assertion;
+    alias_assertion.add(goto_programt::make_assertion(
+      compatible, instruction_iterator->source_location));
+    program.insert_before_swap(instruction_iterator, alias_assertion);
+    ++instruction_iterator;
+  }
```
Comment (feliperodri) [inline src/goto-instrument/code_contracts.cpp]: You just moved this code, correct?

File: src/goto-instrument/code_contracts.h [hunk 20]
```diff
@@ -165,6 +166,13 @@ class code_contractst
 
   void
   add_contract_check(const irep_idt &, const irep_idt &, goto_programt &dest);
+
+  // If the expression is a quantified expression, this function adds
+  // the quantified variable to the symbol table and to the expression map
```
Comment (feliperodri) [inline src/goto-instrument/code_contracts.h]: Keep comment style for methods.
```cpp
  /// If the expression is a quantified expression, this function adds
  /// the quantified variable to the symbol table and to the expression map
```

**Review by feliperodri — CHANGES_REQUESTED** [review]

File: regression/contracts/quantifiers-exists-ensures-01/main.c (hunk 2 above)
Comment (SaswatPadhi) [reply regression/contracts/quantifiers-exists-ensures-01/main.c]: Yes. `forall i . (0 <= i && i < 10) && arr[i] == i` is simply false. It (and similar predicates across all regression tests) should use implication.

Comment (ArenBabikian) [reply regression/contracts/quantifiers-exists-ensures-01/main.c]: I am using `(0 <= i && 1 < 10) && arr[i] == i` within an existential quantifier. I believe that it is correct as is.

Comment (SaswatPadhi) [reply regression/contracts/quantifiers-exists-ensures-01/main.c]: The correctness would depend on what we want to check about the function. Is it:
1. If there is some value within the range then the property holds, or
2. There must be a value within the range _and_ the property should hold

Consider this example:
```
exists x : char . 1000 <= (int) x <= 5000 ==> pred(x)
```
This would evaluate to `true` and models (1) correctly because there is no char between `1000` and `5000` in the first place.

```
exists x : char . 1000 <= (int) x <= 5000 && pred(x)
```
This would evaluate to `false` and models (2) correctly.

File: src/goto-instrument/code_contracts.cpp (hunk 19 above)
Comment (ArenBabikian) [reply src/goto-instrument/code_contracts.cpp]: Yes, I had just fixed the indentation, but I will get rid of these changes as well

File: regression/contracts/function_check_02/test.desc (hunk 1 above)
Comment (ArenBabikian) [reply regression/contracts/function_check_02/test.desc:3]: Sory, it is no longer a known bug, I will adjust the first line.

File: src/goto-instrument/code_contracts.cpp (hunk 15 above)
Comment (ArenBabikian) [reply src/goto-instrument/code_contracts.cpp:180]: I think using an INVARIANT would restrict this function too much, especially down the line, when this function will become recursive. In my current design, I call this function on every function contract, regardless whether they contain a quantified expression or not. If no quantified expression is found, then this function simply does nothing. However, having an INVARIANT would cause the program to fail if no quantifier is found, which is not what desired behavioor, in my opinion.

**Comment by ArenBabikian** [comment]
I have assess all code review comments with the exception of [this comment](https://github.com/diffblue/cbmc/pull/5959/files/160d8873c7f842afdf533d4448f428d48a6f36f7#diff-1ff6abca57de19a68456666ed5c25ce09fcaae20fdc9c2b109cb60cb311c183a) about using `INVARIANT()` for handling  quantifiers (I believe that in the medium-to-long-term, this would not be useful).

File: regression/contracts/quantifiers-exists-ensures-01/test.desc [hunk 21]
```diff
@@ -0,0 +1,12 @@
+CORE
+main.c
+--enforce-all-contracts
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+--
+--
+The purpose of this test is to ensure that we can safety use __CPROVER_exists
```
Comment (feliperodri) [inline regression/contracts/quantifiers-exists-ensures-01/test.desc]: `s/safety/safely/g`

File: regression/contracts/quantifiers-exists-ensures-02/test.desc [hunk 22]
```diff
@@ -0,0 +1,19 @@
+KNOWNBUG
+main.c
+--enforce-all-contracts
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
+--
+The purpose of this test is to ensure that we can safety use __CPROVER_exists
```
Comment (feliperodri) [inline regression/contracts/quantifiers-exists-ensures-02/test.desc]: `s/safety/safely/g`

File: regression/contracts/quantifiers-exists-requires-01/test.desc [hunk 23]
```diff
@@ -0,0 +1,12 @@
+CORE
+main.c
+--replace-all-calls-with-contracts
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+--
+--
+The purpose of this test is to ensure that we can safety use __CPROVER_exists
```
Comment (feliperodri) [inline regression/contracts/quantifiers-exists-requires-01/test.desc]: `s/safety/safely/g`

File: regression/contracts/quantifiers-exists-requires-02/test.desc [hunk 24]
```diff
@@ -0,0 +1,19 @@
+KNOWNBUG
+main.c
+--replace-all-calls-with-contracts
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
+--
+The purpose of this test is to ensure that we can safety use __CPROVER_exists
```
Comment (feliperodri) [inline regression/contracts/quantifiers-exists-requires-02/test.desc]: `s/safety/safely/g`

File: regression/contracts/quantifiers-forall-both-01/test.desc [hunk 25]
```diff
@@ -0,0 +1,12 @@
+CORE
+main.c
+--enforce-all-contracts
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+--
+--
+The purpose of this test is to ensure that we can safely use __CPROVER_forall
+in a negative context. By using the --enforce-all-contracts flag, this test
```
Comment (feliperodri) [inline regression/contracts/quantifiers-forall-both-01/test.desc]: We cannot *always* safely use `__CPROVER_forall` in negative contexts, correct? I'd just add this here too. Same for `__CPROVER_exists` test cases.

File: regression/contracts/quantifiers-forall-ensures-01/test.desc [hunk 26]
```diff
@@ -0,0 +1,12 @@
+CORE
+main.c
+--enforce-all-contracts
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+--
+--
+The purpose of this test is to ensure that we can safety use __CPROVER_forall
```
Comment (feliperodri) [inline regression/contracts/quantifiers-forall-ensures-01/test.desc]: `s/safety/safely/g`

File: regression/contracts/quantifiers-forall-ensures-02/test.desc [hunk 27]
```diff
@@ -0,0 +1,12 @@
+CORE
+main.c
+--enforce-all-contracts
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
+--
+The purpose of this test is to ensure that we can safety use __CPROVER_forall
```
Comment (feliperodri) [inline regression/contracts/quantifiers-forall-ensures-02/test.desc]: `s/safety/safely/g`

File: regression/contracts/quantifiers-forall-requires-01/test.desc [hunk 28]
```diff
@@ -0,0 +1,12 @@
+CORE
+main.c
+--replace-all-calls-with-contracts
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+--
+--
+The purpose of this test is to ensure that we can safety use __CPROVER_forall
```
Comment (feliperodri) [inline regression/contracts/quantifiers-forall-requires-01/test.desc]: `s/safety/safely/g`

File: regression/contracts/quantifiers-forall-requires-02/test.desc [hunk 29]
```diff
@@ -0,0 +1,12 @@
+CORE
+main.c
+--replace-all-calls-with-contracts
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+--
+--
+The purpose of this test is to ensure that we can safety use __CPROVER_forall
```
Comment (feliperodri) [inline regression/contracts/quantifiers-forall-requires-02/test.desc]: `s/safety/safely/g`

File: src/goto-instrument/code_contracts.cpp (hunk 14 above)
Comment (feliperodri) [reply src/goto-instrument/code_contracts.cpp:177]: Could you tag here the GitHub issue to track this comment?

**Review by feliperodri — APPROVED** [review]
LGTM, I only spotted a few more typos and minor comments. Thank you for implementing all comments!!

File: regression/contracts/quantifiers-exists-both-02/test.desc [hunk 30]
```diff
@@ -0,0 +1,19 @@
...
+in a negative context (which is currently not always the case). By using the
+--enforce-all-contracts flag, this test assumes the statement in 
+__CPROVER_requires, then asserts the same statement (in __CPROVER_ensures),
+thus, verification should be successful.
+
+Known bug:
+The current implementation cannot handle a structure such as
+__CPROVER_assume(__CPROVER_exists(int i; pred(i))), where i is
+not explicitly bounded to a predefined range (i.e. if at least
+one of its bound is only declared and not defined).
```
Comment (martin-cs) [inline regression/contracts/quantifiers-exists-both-02/test.desc:19]: Have you tried the SMT back-end?

File: src/goto-instrument/code_contracts.cpp (hunk 14 above)
Comment (martin-cs) [reply src/goto-instrument/code_contracts.cpp:177]: I don't think it should be an `INVARIANT` because it can be triggered by user input.  Some kind of warning / error and an issue would be amazing.

**Review by martin-cs — APPROVED** [review]
Looks good.  In the perfect world, the code changes and the regression tests would be in separate commits but don't worry about changing that unless you have to push a new version.

File: regression/contracts/quantifiers-forall-both-02/main.c [hunk 31]
```diff
@@ -0,0 +1,18 @@
+// clang-format off
+int f1(int *arr, int *len) __CPROVER_requires(__CPROVER_forall {
+  int i;
+  (0 <= i && i < len) ==> arr[i] == 0
```
Comment (SaswatPadhi) [inline regression/contracts/quantifiers-forall-both-02/main.c:4]: This looks incorrect.

Did you mean `int len` in place of `int *len` in the function signature?

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]

File: regression/contracts/quantifiers-forall-both-02/main.c [hunk 32]
```diff
@@ -0,0 +1,18 @@
...
+})
+// clang-format on
+{
+  return 0;
+}
+
+int main()
+{
+  int len, arr[8];
```
Comment (SaswatPadhi) [inline regression/contracts/quantifiers-forall-both-02/main.c:16]: I think `len` should also be initialized (to 8).

Comment (ArenBabikian) [reply regression/contracts/quantifiers-forall-both-02/main.c:16]: Such a case is already tested for in the `quantifiers-forall-both-01` test, and it works with the current implementation. The purpose of this test is in fact to check wether CBMC can handle cases where the range of the quantified variable is not explicitly set a priori.

**Review by SaswatPadhi — APPROVED** [review]

File: regression/contracts/quantifiers-exists-ensures-01/main.c (hunk 2 above)
Comment (SaswatPadhi) [reply regression/contracts/quantifiers-exists-ensures-01/main.c]: @ArenBabikian your original change above was fine in this case.

Check (1) above actually checks: there exists a value such that if it's in the range, then it satisfies property.

The proper way to handle the "if there is a value" case would be to add an implication for non-emptiness:
https://github.com/diffblue/cbmc/blob/66d9d175919787fb08107fe4d28759f284be8f46/regression/contracts/quantifiers-loop-03/main.c#L17-L24

Sorry about the confusion; fixing this (among other issues) in #6145.
